src/enumerable.cr

```diff
@@ -609,6 +609,37 @@ module Enumerable(T)
     memo
   end
 
+  # Similar to `reduce`, but instead of raising when the input is empty,
+  # return `nil`
+  #
+  # ```
+  # ([] of Int32).reduce? { |acc, i| acc + i } # => nil
+  # ```
+  def reduce?
+    memo = uninitialized T
```

Comment:
This and `reduce` should be refactored to a `private def reduce_impl` which returns `{memo, found}`.

Reply:
On second thought this would be hard to do, this is fine as-is.